src/controllers/http/worker-http-controller.ts

```diff
@@ -0,0 +1,191 @@
+
+import { Worker } from "bullmq";
+import { Redis, Cluster } from "ioredis";
+import { debug } from "../../utils/log";
+
+import { HttpHandlerOpts, WorkerMetadata } from "../../interfaces";
+import { validateWorkerMetadata } from "../../validators";
+import { config } from "../../config";
+
+const debugEnabled = config.debugEnabled;
+
+const workers: { [queueName: string]: Worker } = {};
+
+const workerMetadataKey = config.workerMetadataKey;
+
+// Gracefully close all workers
+process.on('exit', async () => {
+  for (const queueName in workers) {
+    await workers[queueName].close();
+  }
+});
+
+const workerFromMetadata = (queueName: string, workerMetadata: WorkerMetadata, connection: Redis | Cluster): Worker => {
+  const { endpoint: workerEndpoint, opts: workerOptions } = workerMetadata;
+
+  debugEnabled && debug(`Starting worker for queue ${queueName} with endpoint ${workerMetadata.endpoint.url} and options ${workerMetadata.opts || 'default'}`);
+
+  const worker = new Worker(queueName, async (job) => {
+    debugEnabled && debug(`Processing job ${job.id} from queue ${queueName} with endpoint ${workerEndpoint.url}`);
+
+    // Process job by calling an external service using the worker endpoint
+    const controller = new AbortController()
+    const timeoutId = setTimeout(() => {
+      debugEnabled && ("Timeout, aborting...")
+      controller.abort()
+    }, workerEndpoint.timeout || 3000)
+
+    try {
+      const response = await fetch(workerEndpoint.url, {
+        method: workerEndpoint.method,
+        headers: workerEndpoint.headers,
+        body: JSON.stringify(job.toJSON()),
+        signal: controller.signal
+      });
+
+      if (!response.ok) {
+        throw new Error(response.statusText);
+      }
+      debugEnabled && debug(`Job "${job.id}:${job.name}" processed successfully`);
+
+      // If response is supposed to be json lets parse it
+      if (response.headers.get('content-type')?.includes('application/json')) {
+        return response.json();
+      } else {
+        return response.text();
+      }
+    } catch (err) {
+      debugEnabled && debug(`Failed to process job ${job.id}: ${(<Error>err).message}`);
+      throw err;
+    } finally {
+      clearTimeout(timeoutId);
+    }
+  }, { ...workerOptions, connection, prefix: config.defaultQueuePrefix });
+
+  if (debugEnabled) {
+    worker.on('failed', (job, err) => {
+      debug(`Job ${job!.id} failed: ${err}`);
+    });
+  }
+
+  return worker;
+};
+
+export const WorkerHttpController = {
+  init: (redisClient: Redis | Cluster) => {
+    // Load workers from Redis and start them
+    debugEnabled && debug('Loading workers from Redis...');
+    const stream = redisClient.hscanStream(workerMetadataKey, { count: 10 });
```

Comment:
Maybe something similar to what we used for queue discovery could be helpful here too:

```
public static async getActiveQueueIds(redis: Redis): Promise<string[]> {
		await redis.zremrangebyscore(
			QUEUE_IDS_KEY,
			'-inf',
			Date.now() - daysToMilliseconds(RETENTION_QUEUE_IDS_IN_DAYS),
		)
		const queueIds = await redis.zrange(QUEUE_IDS_KEY, 0, -1)
		return queueIds.sort()
	}

	public async start(): Promise<void> {
		await this.redis.zadd(QUEUE_IDS_KEY, Date.now(), this.config.queueId)
}
```

Reply:
For the proxy we need to store a complete json object for every queue, so I don't know if ZSET is a good structure for that since it is a nice property of hsets to be able to update a workers options without needing to iterate through all the zset items.

Reply:
Would hscan work well on a big Redis store? Wouldn't it iterate over the whole thing? maybe some key lookup map would be helpful?

Reply:
It would only iterate on the given hashmap, and since it is scanning it is not keeping redis busy. Also, this operation is only needed when restarting the proxy. The number of workers should not be very big either...